### PR TITLE
Extract zk-file-name-regexp

### PR DESCRIPTION
**Pull request by boyechko**
This regexp is created several times thruoughout the code, so refactoring it
like this will make any future changes easier and less error-prone.

P.S. Flycheck is magical! Thanks for the tip.

### CODE REVIEW

**Comment by boyechko** [comment]
There is a very minor cost added for using `(zk-file-name-regexp)` rather than a variable, but it only becomes apparent at very large number of runs (100,000). For moderate number of notes, it should not be a problem:

``` emacs-lisp
(benchmark-run 10000
  (let ((file "~/Zk/standard/202111102331 {Talk} Animal History in the Anthropocene.txt")
        (zk-id-regexp "\\([0-9]\\{12\\}\\)"))
    (when (string-match (zk-file-name-regexp) file)
      (match-string-no-properties 2 file))))
(0.029169 0 0.0)

(benchmark-run-compiled 10000
  (let ((file "~/Zk/standard/202111102331 {Talk} Animal History in the Anthropocene.txt")
        (zk-id-regexp "\\([0-9]\\{12\\}\\)"))
    (when (string-match zk--file-name-regexp file)
      (match-string-no-properties 2 file))))
(0.026916 0 0.0)
```

NB: I temporarily defined `zk--file-name-regexp` variable with the same code as `(zk-file-name-regexp)`. By the way, the function is public (as opposed to `zk--`) because it might be useful for users to locate the right files.

File: zk.el [hunk 1]
```diff
@@ -309,6 +296,31 @@ otherwise just match against `zk-file-name-regexp'."
...
+          "."
+          "\\(?2:.*?\\)"
+          "\\."
+          zk-file-extension
+          ".*"))
+
+(defun zk--file-name-id (file)
+  "Return the ID of the given FILE.
+This relies on match data from `zk-file-name-regexp`."
+  (match-string-no-properties 1 file))
```
Comment (localauthor) [inline zk.el:315]: I don't think these instances of `match-string` should be extracted. Since they rely on match data, they are contingent functions and don't have any use independent of very specific conditions (ie, following an instance where there is the appropriate match data provided by `zk-file-name-regexp`.

File: zk.el [hunk 2]
```diff
@@ -191,10 +191,10 @@ Used in conjunction with `format', the string `%s' will be
 replaced by a note's ID."
   :type 'string)
 
-;; This needs to be a macro in order to reflect user changes to the variables.
-(defmacro zk-link-regexp ()
-  "Returns regexp for zk-link based on `zk-link-format' and `zk-id-regexp'."
-  '(format (regexp-quote zk-link-format) zk-id-regexp))
+(defun zk-link-regexp ()
```
Comment (localauthor) [inline zk.el:194]: It's not clear to me what difference it makes if this is a function vs a macro. Do you know the implications here?

File: zk.el [hunk 3]
```diff
@@ -278,19 +278,48 @@ Adds zk-id as an Embark target, and adds `zk-id-map' and
...
     (and file
-         (file-exists-p file)
-         (string-match-p zk-id-regexp file)
+         (string-match (zk-file-name-regexp) file)
          (or (not strict)
-             (file-in-directory-p file zk-directory)))))
+             (save-match-data
+               (file-in-directory-p file zk-directory))))))
+
+(defun zk--de-separator (string)
```
Comment (localauthor) [inline zk.el:295]: I don't think we need this function, since it doesn't save many lines, and it only obscures (abstracts) what is happening in the functions it appears it. Saving a few lines isn't a huge priority if it makes things less clear.

File: zk.el (hunk 2 above)
Comment (boyechko) [reply zk.el:194]: The implication is that a macro generates code that is then simply inserted wherever the macro is called from, so there is no run-time performance difference at all between using a macro and just copy/pasting the code it generates. However, if the defmacro definition is re-evaluated, the code that used that macro will not be updated until it is compiled or evaluated again:

``` emacs-lisp
(defmacro foo (bar)
  "Foo that bar."
  `(list 'fooing 'that ,bar))
(foo 'lamp) ;; => (fooing that lamp)
(defun foobar (baz)
  "Foo the bar with baz."
  (foo baz))
(foobar 'cup) ;; => (fooing that cup)
(defmacro foo (bar)
  "Foo that bar differently."
  `(list 'fooing 'that ,bar 'differently))
(foo 'lamp) ;; => (fooing that lamp differently)
(foobar 'book) ;; => (fooing that book)
```

So, not really a big deal for end use, but can make coding and debugging a pain if you forget to re-evaluate the functions that use the macro that you changed.

File: zk.el (hunk 1 above)
Comment (boyechko) [reply zk.el:315]: I see your point, but creating accessors like this allows to abstract the meaning of the code away from implementation. If you change what groups 1 and 2 mean, for example, then you will need to find all instances of `(match-string[-no-properties]` and update them. Having the accessors in place means that the code will need to be changed in one place only (i.e. accessors).

Also, the next step in refactoring might be to put all the string matching into the functions themselves, so the whole construct of
``` emacs-lisp
(when (match-string (zk-file-name-regexp) file)
  (replace-regexp-in-string zk-file-name-separator
                            " "
                            (match-string-no-properties 2 file))
```
would become just `(zk-file-name-title file)`. At that point, one could rewrite `zk--parse-file` to the following:

``` emacs-lisp
(defun zk--parse-file (target files)
  (let ((result
         (mapcar (pcase target
             ('id #'zk--file-name-id)
             ('title #'zk--file-name-title)
             (_
               (signal 'wrong-type-argument '(target (or 'id 'title)))))
            (if (listp files)
                files
              (list files)))))
    (if (cdr result)
        result
      (car result))))
```

In other words, introducing a layer of abstraction for raw data has benefits, I feel.

Comment (localauthor) [reply zk.el:315]: Under what circumstances would the meaning of groups 1 and 2 be changed? Their meaning as ID and FILE is pretty much essential, right?

As they are currently, `zk--file-name-id` and `-title` are almost just aliases for `match-string`.

Perhaps they would be easier to justify if they did the string matching themselves, but that would mean that each function would run `string-match` twice (once for id and once for title) instead of just once. So, it's kind of splitting hairs, no?

Comment (boyechko) [reply zk.el:315]: True, it's hard to imagine them changing. So, yes, they are definitely just aliases for `match-string`. I suppose at the core of it not liking "magical numbers," and `(zk--file-name-title file)` is more readable than `(match-string 1 file)`. If only Emacs regular expression syntax had named groups like other modern languages rather than just merely numbered ones (so one could write regexps like `\\(?<id>[0-9]\\{12\\}\\).\\(?<title>.*\\)`). Maybe one day.

I get your point about doing `string-match` twice. The performance should not be affected much, but it's your package, so if you prefer to keep that part as is, I can `git revert` the relevant commits, just keep the `zk-file-name-regexp` part.

Comment (boyechko) [reply zk.el:315]: I was curious about performance, so after moving `string-match` into the `zk--file-name-id` and `zk--file-name-title`, I instrumented all `zk-` functions with `elp.el` and spent half an hour or so "conversing" with my Zettelkasten. The results are not bad, but the added total of three seconds (compared to 8 seconds for `zk-index-search') is probably not worth it, indeed.

[zk--file-name-title and zk--file-name-id take up 3 seconds for 100K calls each, zk-file-name-regexp just a half-second despite 350K calls.](https://postimg.cc/xXFNj1jW "zk--file-name-title and zk--file-name-id take up 3 seconds for 100K calls each, zk-file-name-regexp just a half-second despite 350K calls.")

**Comment by localauthor** [comment]
This looks good to me. Is it ready to merge?

**Comment by boyechko** [comment]
> This looks good to me. Is it ready to merge?

Yes, everything seems to work as expected.
